src/plugin/src/main/java/com/defenseunicorns/uds/keycloak/plugin/authentication/RequireGroupAuthenticator.java

```diff
@@ -0,0 +1,148 @@
+            return;
+        }
+
+        handleGroupAuthentication(context, user, groupsNode);
+    }
+
+    private JsonNode parseGroupsAttribute(String groupsAttribute) {
+        ObjectMapper objectMapper = new ObjectMapper();
+        try {
+            return objectMapper.readTree(groupsAttribute).path("anyOf");
+        } catch (Exception e) {
+            LOGGER.errorf("Failed to parse groups JSON: %s", e.getMessage());
+            return null;
+        }
+    }
+
+    private void handleGroupAuthentication(AuthenticationFlowContext context, UserModel user, JsonNode groupsNode) {
+        if (!groupsNode.isArray() || groupsNode.size() == 0) {
+            logAndFail(context, "Groups attribute does not contain a valid anyOf array", AuthenticationFlowError.INVALID_CLIENT_SESSION);
+            return;
+        }
+
+        RealmModel realm = context.getRealm();
+        boolean foundGroup = false;
+
+        for (JsonNode groupNode : groupsNode) {
+            String groupName = groupNode.asText();
+            Optional<GroupModel> group = getGroupByPath(groupName, realm);
+
+            if (group.isPresent()) {
+                if (isMemberOfExactGroup(user, group.get())) {
+                    LOGGER.infof("User %s is authorized for group %s", user.getUsername(), groupName);
+                    context.success();
+                    return;
+                } else {
+                    foundGroup = true;
+                }
+            }
+        }
+
+        if (foundGroup) {
+            logAndFail(context, "User is not a member of the required group(s)", AuthenticationFlowError.INVALID_CLIENT_SESSION);
+        } else {
+            logAndFail(context, "Required group(s) do not exist in realm", AuthenticationFlowError.INVALID_CLIENT_SESSION);
+        }
+    }
+
+    private Optional<GroupModel> getGroupByPath(final String groupPath, final RealmModel realm) {
+        return realm.getGroupsStream()
+            .filter(group -> buildGroupPath(group).equals(groupPath))
+            .findFirst();
+    }
+
+    private String buildGroupPath(GroupModel group) {
+        StringBuilder path = new StringBuilder();
+        GroupModel currentGroup = group;
+        while (currentGroup != null) {
+            path.insert(0, "/" + currentGroup.getName());
+            currentGroup = currentGroup.getParent();
+        }
+        return path.toString();
+    }
+
+    private boolean isMemberOfExactGroup(UserModel user, GroupModel group) {
+        return user.getGroupsStream()
+            .anyMatch(userGroup -> buildGroupPath(userGroup).equals(buildGroupPath(group)));
+    }
+
+    private void logAndFail(AuthenticationFlowContext context, String message, AuthenticationFlowError error) {
+        LOGGER.warn(message);
+        context.failure(error);
+    }
+
+    @Override
+    public void action(final AuthenticationFlowContext authenticationFlowContext) {
+        // no implementation needed here
+    }
```

Comment:
Can we figure out when a `action()` on an `Authenticator` is called? What does "Called from a form action invocation."

If leaving this empty is correct, that's fine, but let's verify.

If we're sure that this should never be called, then maybe throw an `UnsupportedOperationException`.

Reply:
Based on ChatGPT, the action is called during the authentication flow after the user submits their credentials or interacts with the authentication form. It allows you to handle the user's input and decide the next step in the authentication process.

Essentially, it's a way to define a required action "like" actions that are specific to this authenticator. 

[Some example uses of the authenticator action method](https://github.com/thomasdarimont/keycloak-extension-playground/blob/2e16cb16b85d2ea61a1f6b095e3b01a13ec4af9d/dynamic-email-code-auth-extension/src/main/java/com/github/thomasdarimont/keycloak/auth/emailcode/EmailCodeAuthenticatorForm.java#L72)

So i am fairly confident that leaving this empty is the correct path forward.

Reply:
Hmmm ok. So may be like, if you had a "custom MFA" authenticator you could accept the OTP in a form dialog and this code might handle it? yeah, if that's accurate we don't need this